Reworking the "Riverpod for Provider users" documentation section

lucavenir opened the pull request:
Howdies,

It's been a while since I promised "a great deal of contributions" onto the documentation... which I never did, in the end.

But not this time! I've tried my best to rework this page and address #2629.
I must say, the fact that you carefully wrote exactly what you've wanted helped _a lot_.

I pretty much re-structured the page as follows:

1. Provider's limitations (an in-detail view of Provider's problems and why you won't find them in Riverpod)
2. Towards Riverpod (how was Riverpod born, what relationship there is with Provider, what to choose)
3. Migrating tips (as requested in the linked issue)
4. Differences and similarities between Riverpod and Provider (which also gives migration tips under the hood)
5. (I added a Scoping vs `.family.autoDispose` paragraph in this chapter)

Since I have all sort of insecurities about this, I'm expecting some sort of criticism. Let me know. Overall, I feel like this is too much content to fit into a page. I'd split this into multiple pages, one for the chapters that are now seen there

rrousselGit commented on website/docs/riverpod_for_provider_users.mdx:
```diff
@@ -18,62 +18,179 @@ import {
 wants to learn about Riverpod.
 
-## The relationship between Riverpod and Provider
+Since Provider is widely popular, why would one migrate to Riverpod?
```
That probably should be a heading instead of the current "Provider's limitations"

rrousselGit commented:
Thanks for this! I'll look more into this tomorrow to check everything is fine. I have a few other PRs to review first :)

lucavenir commented:
Perfect, take your time, I am available to implement any correction

saltedpotatos commented on website/docs/riverpod_for_provider_users.mdx:
```diff
@@ -18,62 +18,179 @@ import {
 
-So to answer the question you're probably asking yourself:  
-**Should I use Provider or Riverpod?**
+Instead, this should be taken as a furhter step forward.
```
Small typo

rrousselGit replied:
```suggestion
Instead, this should be taken as a further step forward.
```

rrousselGit commented:
The page looks quite dense with those changes. There's a huge wall of text.

I wonder how it could be made more readable.

rrousselGit commented on website/docs/riverpod_for_provider_users.mdx:
```diff
@@ -18,62 +18,179 @@ import {
-## The relationship between Riverpod and Provider
+Since Provider is widely popular, why would one migrate to Riverpod?
+
+## Provider's limitations
```
Docusaurus already enables writing "articles". I've never had the need for this. But this could be a good use-case

lucavenir replied:
I've wrote all of these limitations in response of #2629 (see bullet points). I interpreted those bullet points as "necessary" to fix #2629. Moving "Provider's limitations" into a separate page makes sense to me

rrousselGit replied:
I know. But keep in mind that #2676 is only a quick draft of what should be covered.  
In the end it's quite important to make the page readable. Otherwise people will simply not read it, because they are scared of walls of texts.

I think that's generally Riverpod's issue. There's too much docs :P

lucavenir replied:
I realize this. I'm sure I can be more concise, but at the same time I wrote these subparagraphs while reminding of all the questions we're given into discord daily.

So you're telling me that on one hand folks want documentation, and on the other they don't want to read it 😆 While this is true, it all boils down to split this page as suggested below.

ALSO some examples (code) might alleviate readability. What do you think?

rrousselGit replied:
The content is great, don't worry about that. I'm mainly worried about formatting, which is not something those GitHub issues care about.

Examples would for sure help. Although it could also have the adverse effect (since the page would become an even larger wall of text).

I'd combine examples with your proposal of splitting this into subpages

lucavenir commented:
> The page looks quite dense with those changes. There's a huge wall of text.
> 
> I wonder how it could be made more readable.

Yes, that's why I proposed to split this section into subsection. Each paragraph in there deserves its own subpage. Just like "All Providers" has its own submenu with its subpages. In other words, you'd have a main "Riverpod for Provider users" section and its subsections would be:

- Provider's limitations
- Towards Riverpod
- Migrating
- Differences and similarities

So it can be digested one bit at a time. All in all #2629 kinda requested this content, so I tried to deliver

catalunha commented:
> So you're telling me that on one hand folks want documentation, and on the other they don't want to read it 😆 While this is true, it all boils down to split this page as suggested below.

Congratulations to Remi and everyone from this and other PRs.
I always read and reread the Riverpod documentation. I believe that the best way to publicize/benefit is really a detailed documentation.
When learning Riverpod I made a comparative chart of the old and new syntax and when I was finishing it had already been covered in the topic
https://docs-v2.riverpod.dev/docs/about_code_generation The Syntax.
It turned out really good. Congratulations to the author of those PRs.

But there are many hidden features besides the trivial Riverpod (read/watch/notifier) that are equally important for learning and would get your readers attention.
Such as:
ref.read(xProvider.future);
ref.read(xProvider).maybeWhen();
and many others.

Now I'm studying how to exemplify the application of each one of them. But if someone already has broader knowledge it would be interesting to start any documentation about it.

In the official examples, little is said about these and other resources.

lucavenir commented:
Generally speaking, I do approve of non-trivial, or rather "not-too-much-concise" documentation, but readability is very important. Folks can't just expect an essay, this is technical documentation.

Imo the key is to inject more code snippets, and split "big pages" into sub chapters pages. Let's see what Remi decides, then I'll work into it asap

Kypsis commented:
The topic of good documentation especially one that doesn't only show trivial solutions is pretty much the most requested feature of any state management(ish) library. One of the reasons why Bloc is so popular is probably just that, documentation. They show you how to solve a lot of the tricker parts with managing state. 

Yet Riverpod in my eyes is clearly the superior and the more future facing approach (and more in line with the rest of the greater UI development sphere). It's just annoying and time consuming going through Discord search, GH issues and other secondary and tertiary means of finding info. So no, you can never have too much info. It is mostly always a case of presenting the info in a digestible manner.

lucavenir commented:
Thank you for the inputs. May I suggest this is not the right place? Let's keep this about a PR that actually tries to solve some information. I'll try my best to address any other documentation request in the next 1-2 months. But this is a community effort. Just submit a PR like I did.

rrousselGit commented:
> You'd obtain a main "Riverpod for Provider users" section and its subsections would be:
> 
> * Provider's limitations
> * Towards Riverpod
> * Migrating
> * Differences and similarities

@lucavenir 
I assume you meant that "Riverpod for Provider users" would become a "folder" in the sidebar, and that those sections would become pages?

At the moment, the documentation doesn't allow a non-folder to define subpages. So we'd need to have the introduction somewhere

lucavenir commented:
Yes exactly that's what I meant. There's no problem. The introduction was minimal. I can still write 2 lines and call it "introduction" and that's it.

lucavenir commented:
Ok I'll do it asap. Splitting into multiple pages (and under a single folder), trying to summarize stuff a bit, and implementing some examples / some "info" panels when necessary, so that reading is smoother.

lucavenir commented:
There we go @rrousselGit.

- I added some snippets here and there
- I refactored the whole section with its folder
- I did more readings / small fixes / etc.

Let me know if this is okay for this section.

rrousselGit commented:
Looks like the CI is failing. Could you fix it? 🙏

lucavenir commented:
Weird. My riverpod_generator appends:
```dart
// ignore_for_file: type=lint
```
at the end of each generated file.

I removed them by hand now.

lucavenir commented:
Hey @rrousselGit I'm just leaving this here notifying you that the CI now works in case you've missed it.

rrousselGit commented:
Yep thanks, will look at it soon~
I've had a lot of notifications these past few days :P

rrousselGit commented:
I'll need a few passes through this before making a final review, as this is quite dense.  
But for now, honestly, I'm skeptical about the flow of the document.

Someone looking for a quick explanation of how to migrate would probably have to click through all the documents.
And it feels a bit unexpected to have some pages which take about a minute to read. Using 5 sections might be too much, and maybe we could do it in 2-3


Maybe an alternative would be to have two docs:
- A quick start, which gives a brief overview of how to do the migration with a line or two of explanation/tip.
- A more in-depth page, probably quite lengthy, covering everything

I don't want to overwork you though. Please tell me if I'm being too hard

rrousselGit commented:
Due to the complexity of the work, if you want to iterate with me about the document flow, it's probably fine to just make suggestions in this thread.   
Since for now, my comments are mainly about the flow as a whole, I don't care about minor details. We can view those later.

I wouldn't want you to rewrite everything when I'm not sure what the outcome should be :)

lucavenir commented:
Hello there,

Yes this is taking more than I've expected, I would have loved to end this here but I realize that offering under-the-expectation content might be worse than offering none.

> if you want to iterate with me about the document flow, it's probably fine to just make suggestions in this thread

Well of course I like the content I've just written it 😛 I put a lot of thought on this structure, considering the above comments (folks expected more content), so it's on you to tell me exactly what you'd like better.

But I realize this is about the whole section "flow", so maybe it isn't about the content itself.
Let's start with a new TOC proposal. Let's keep the "Riverpod for Provider Users" expansion-tile-like section. It should then have the following subsection:

1. Quickstart (yet to be written)
2. Provider vs Riverpod (the content I've written up until now as a "deep topic")

What do you think about this? The first subsection will be quick and fast. The second one will have everything one could ask from this topic.

Let me know

rrousselGit commented:
Yeah I knew you'd think this is taking longer than expected, sorry 🙏   
In the end, I'm quite critical of the "flow". I think that's probably the main issue with the current docs website.

Although thinking about it, for a "From x package" page, the flow probably matters less. I'll try to be less critical.

I still think the current sections aren't ideal from the sidebar's point of view.
I'll probably make you rename a few titles too for the same reasons.  

> * Quickstart (yet to be written)
> * Provider vs Riverpod (the content I've written up until now as a "deep topic", under one big article)

That's good, yes.

With maybe a third "Motivation" section, where we would talk about the technical reasons why a new package is made.  
I'd separate it from the two other pages, as this is background information and less so useful for doing the migration.

lucavenir commented:
Ok, that's perfect. Let's try this out.

rrousselGit commented:
To save time, as a first step, we could only keep the title

No need to focus on the body of the articles until the titles/sections are nailed :)

lucavenir commented:
@rrousselGit there we go, this is just a draft of course.

Three sections:

- Quick Start
- Provider VS Riverpod
- Motivation

They have increasing verbosity and in-depth references. They're meant to be this way, so that one could still search for a deep explanation, while having a quick start page available.

Let me know if this is the right direction so I can start arranging the content (summarizing it, expanding it, adding code, etc.).

rrousselGit commented on website/docs/from_provider/quickstart.mdx:
```diff
@@ -0,0 +1,115 @@
+Migrating from Provider to Riverpod can be very straightforward.  
+Migrating basically consists in a few steps that can be done in an *incremental* way.
+
+### Start gently
```
As a title, it's not super clear I think.
I'd consider:

```suggestion
### Start with ChangeNotifierProvider
```


And the content likely would talk about how Riverpod voluntarily has ChangeNotifierProvider for easy migration.

Then we'd probably want to come back to this page to cross-link the StateNotifier->Notifier migration guide once it's out there. Let me update the draft for it

rrousselGit replied:
We'd probably want a very quick note about autoDispose here too.
Something along the lines of "This applies to autoDispose. Although it is preferred to enable it, you can do that at a later stage".

rrousselGit commented on website/docs/from_provider/quickstart.mdx:
```diff
@@ -0,0 +1,115 @@
+:::
+
+### Migrate one Provider at a time
+*All Providers from pkg:provider have a strict equivalent in Riverpod*.
```
Technically no if you consider ProxyProvider variants

rrousselGit replied:
We're missing that migration btw.

The migration involves using ref.listen

lucavenir replied:
> The migration involves using ref.listen

Could you elaborate on this? Is mentioning the fact that Providers in pkg:Riverpod are composable by default enough? Or do you have something else in mind? (:

lucavenir replied:
> We're missing that migration btw.

I've added a small paragraph in the quickstart, cross referencing the "deeper" article. Let me know if that's what you intended.

rrousselGit commented on website/docs/from_provider/quickstart.mdx:
```diff
@@ -0,0 +1,115 @@
+Thus, migrating consists in moving `provider.Provider` definitions *outside* of `MultiProvider`, 
+defining it as a final global `riverpod.Provider` variable and using `riverpod`'s new classes where needed.
+
+
```
Another section that would be valuable to add here would be "Start with _leaves_"

Aka start with providers who do not depend on anything else. So avoid ProxyProviders first basically.
And only start migrating them once all of their dependencies have been migrated.

rrousselGit replied:
We probably should also talk about the "eager" flag on providers

As in, providers in Riverpod are always lazy. If you want them to be eagerly initialized, you need to read the provider in the place where you used to have your MultiProvider

rrousselGit commented on website/docs/from_provider/provider_vs_riverpod.mdx:
```diff
@@ -191,7 +131,7 @@ Inside the `build` method, use "watch". Inside click handlers and other events,
 :::
 
-### Reading providers: Consumer
+## Reading providers: Consumer
```
I think it's be worth talking about Consumer2 vs calling ref.watch multiple times

rrousselGit commented on website/docs/from_provider/provider_vs_riverpod.mdx:
```diff
@@ -191,7 +131,7 @@ Inside the `build` method, use "watch". Inside click handlers and other events,
 use "read".
 :::
 
```
We may want a section talking about context.select -> ref.watch(provider.select)

lucavenir commented:
I've read your comments. Let's summarize it here:

- [ ] Change a title to "Start with `ChangeNotifierProvider`"
- [ ] Cross reference, or link, the article about "`StateNotifier` to `Notifier` migration" (see note below)
- [ ] Specify that `ProxyProvider` has not direct equivalent, except Riverpod's providers can be chained by default (I'm not sure what you mean with `The migration involves using ref.listen` tho)
- [ ] Move the "alias provider and riverpod libs" section in the right place as shown above
- [ ] Add a "Start with leaves" section in the quick start as suggested above
- [ ] Add a "You don't need to use Consumers ASAP" section in the quickstart as suggested above
- [ ] Mention that Provider's `eager` flag isn't a thing in Riverpod and show the difference (I'd suggest to move this suggestion in the "Riverpod vs Provider" section though, i.e. not in the quickstart)

@rrousselGit I just need your green light and I'll address these ASAP.

### Stuff I didn't understand yet
> I think it's be worth talking about Consumer2 vs calling ref.watch multiple times

I'm not sure what you mean here

> We may want a section talking about context.select -> ref.watch(provider.select)

Same here


## Note about the "from `StateNotifier` to `Notifier`" article
Does this page exist yet? Or is it still needed to address this topic (aka has anyone started a draft on this)?

If it's there, could you point me towards the issue mentioning it?

If it is not there, I'd gladly address this missing piece of the documentation. To my understanding this should be called "from 1.0 to 2.0", similarly to [this article](https://riverpod.dev/docs/migration/0.14.0_to_1.0.0). If anything, I'd put this topic right below that.

rrousselGit commented:
> @rrousselGit I just need your green light so that I can address these ASAP.



Yep looks good. We can talk about those individual points in the associated comments to help the disucussion

> Does this page exist yet? Or is it still needed to address this topic (aka has anyone started a draft on this)?



Nope. It's a different issue. That comment was mainly for me/context
This is the issue: https://github.com/rrousselGit/riverpod/issues/2630




> > I think it's be worth talking about Consumer2 vs calling ref.watch multiple times
> 
> I'm not sure what you mean here

pkg:provider has Consumer vs Consumer2 vs Consumer3, with more generics.
Whereas Riverpod has a single Consumer. But we instead use `ref.watch` multiple times in the widget

Users of Consumer2 may want to know how to migrate to Riverpod





> > We may want a section talking about context.select -> ref.watch(provider.select)
> 
> Same here

It's about mentioning how both Riverpod & Provider have the "select" functionality. But the syntax is slightly different, so we need to guide users on how to migrate.

rrousselGit commented:
And again, please tell me if I'm asking too much.

I don't mind continuing the work myself if you want to stop. But of course, I love the help.

AhmedLSayed9 commented:
This PR is great.
It will greatly encourage and help many folks in migrating from Provider to Riverpod.

lucavenir commented:
Ok, looks cool up until now.

Now I just need to cross reference the `StateNotifier` migration and we're good to go.
This means that this PR is blocked by #2630 (see my proposal there). No big deal.

Let me know if the current version looks OK for you and let me know what you want to improve (once again).

rrousselGit commented:
I'm working on reviewing all docs pages atm, so I have yet to look at your updates. But I just thought:

It may be worth telling users how to migrate ChangeNotifierProvider to codegen


Folks can do:

```dart
extension on Ref {
  T listenAndDisposeChangeNotifier<T extends ChangeNotifier>(T notifier) {
    notifier.addListener(notifyListeners);
    onDispose(() => notifier.removeListener(notifyListeners);
    onDispose(notifier.dispose);
    return notifier;
  }
}

@riverpod
MyNotifier example(ExampleRef ref) {
  return ref.listenAndDisposeChangeNotifier(MyNotifier());
}
```

I'll likely add this to the code-gen migration page too.


_Back to working on docs_

lucavenir commented:
Sure thing, let me add this in its section. I'm busy this weekend but I'll work on it asap

rrousselGit commented:
No problem.

By the way since I'm trying to publish v2 docs this weekend, I may end-up editing this PR myself.

lucavenir commented:
Hey there,

> It may be worth telling users how to migrate ChangeNotifierProvider to codegen

Done in the last commit.

I've also looked at the `rework-flow` branch. It looks *amazing*, no cap 🧢 
But I am unsure how to rebase this PR with that branch (which doesn't appear in this fork).

Also, apparently I am not allowed to fork a project multiple times (which might make sense tbh).
I don't think deleting this fork is an option either, it feels way too risky 🥶

I am sorry I am not that much of a git expert, any guidance would help.

rrousselGit commented:
Run: 

```
git remote add rrousselGitRemote https://github.com/rrousselGit/riverpod
git pull rrousselGitRemote rework-flow
```

lucavenir commented:
I've just merged `rework-flow` into this fork. I've addressed the sidebar changes.

I've also re-arranged the final paragraphs in "motivations" to be aligned with the new flow.

Let me know what you want next from this PR. If this is okay, I'd switch to something else, such as the `StateNotifier` migration topic.

rrousselGit commented:
Epic, thanks!
